src/diskio/mod.rs

```
@@ -0,0 +1,216 @@
+        let len = content.len();
+        Item {
+            full_path,
+            kind: Kind::File(content),
+            start: 0.0,
+            finish: 0.0,
+            size: Some(len),
+            result: Ok(()),
+            mode,
+        }
+    }
+}
+
+/// Trait object for performing IO. At this point the overhead
+/// of trait invocation is not a bottleneck, but if it becomes
+/// one we could consider an enum variant based approach instead.
+pub trait Executor {
+    /// Perform a single operation.
+    /// During overload situations previously queued items may
+    /// need to be completed before the item is accepted:
+    /// consume the returned iterator.
+    fn execute(&mut self, mut item: Item) -> Box<dyn '_ + Iterator<Item = Item>> {
+        item.start = precise_time_s();
+        self.dispatch(item)
+    }
+
+    /// Actually dispatch a operation.
+    /// This is called by the default execute() implementation and
+    /// should not be called directly.
+    fn dispatch(&mut self, item: Item) -> Box<dyn '_ + Iterator<Item = Item>>;
+
+    /// Wrap up any pending operations and iterate over them.
+    /// All operations submitted before the join will have been
+    /// returned either through ready/complete or join once join
+    /// returns.
+    fn join(&mut self) -> Box<dyn '_ + Iterator<Item = Item>>;
+
+    /// Iterate over completed items.
+    fn completed(&mut self) -> Box<dyn '_ + Iterator<Item = Item>>;
+}
+
+/// Trivial single threaded IO to be used from executors.
+/// (Crazy sophisticated ones can obviously ignore this)
+pub fn perform(item: &mut Item) {
+    // directories: make them, TODO: register with the dir existence cache.
+    // Files, write them.
+    item.result = match item.kind {
+        Kind::Directory => create_dir(&item.full_path),
+        Kind::File(ref contents) => write_file(&item.full_path, &contents, item.mode),
+    };
+    item.finish = precise_time_s();
+}
+
+#[allow(unused_variables)]
```

Rather than unused variables (which I assume is `mode` on Windows) could we not check `mode` against `0644/0755` and return an error if the mode is bad as part of our tar verification plans?

---

Latest rust can do attributes on parameters. Yay. But no, wrong later. push back here. I can call it _mode instead if you prefer to have it narrow; alex didn't like that in tar-rs.

---

I'm fine for it to stay like this for now.